### ISSUE TITLE
Proposition of high-level API for libnitrokey [draft]

### ISSUE THREAD

**Issue by szszszsz** [Nitrokey/libnitrokey#5]
Please comment what functions should have high-level API to make libnitrokey easy to use for user or to integrate into other frameworks. Please propose a name, parameters, return value (if needed) and description what should it do. This could be later used as reference.

Example:
int32_t getHOTPCode(int slotNumber) - returns HOTP code for current counter for slot configured before, throws or returns <0 if slot is not programmed.

**Comment by cornelinux** [comment]
I would like to start with initializing OTP slots. Thus I probably would need stuff like this:

Login
====
I probably need to login as the user or SO to be able to programm the OTP part of the nitrokey.

    int NK_login( user, pin, c )

would either login as User or SO.
*c* being the number of the connected NK, if more than one connected NK is supported.
Returns some sensible error/return code and/or a session handle.

Logout
=====
The session to the nitrokey gets discarded.

    int NK_logout( session_handle )

With >=0 in case of success.

List Slots
=======

I want to know, which slots are already filled. 

     some_struct NK_list_slots( session_handle )

The structure can contain necessary information.
I at least need to know the name of the slot and if it is occupied.

Erase Slot
========
We might want to erase a slot (if we need to erase it before we can write it)

    int NK_erase_slot(session_handle, slot_num)

Erases the slot number *slot_num*.
Returns return_code.

Get Random number
================

I think the NK has a slightly better RNG than most normal systems? So it would be great if we could use the RNG to create the H/TOTP secret.
Hm, so in fact we need a random string. Or hex string....

Write Slot
========
This is obviously the most important part. I would like to be able to write a TOTP and HOTP slot. If it is sensible to have one function or two, I do not know.

    int NK_write_hotp_slot(session_handle, secret, hotp_counter)

*secret* being the HOTP secret key (does the NK only support HMAC-SHA1 or also HMAC-SHA256? - then we would need to add an additional parameter)
*hotp_counter* being the initial counter. Defaults to 0.
If these are only a few parameters, I would prefer normal parameters instead of passing a complicates structure as parameter. (I will have to write a python wrapper for this...)

The corresponding totp-function would have additional parameters:

    int NK_write_totp_slot(session_handle, secret, time_window)

*time_window* being the 30 or 60 seconds.

To me this feels like Chrismas. Great ideas and wishes but having some slight doubt if santa will brings all this this year ;-)

**Comment by szszszsz** [comment]
Hi @cornelinux !
Santa is earlier this year :-) Please check latest commit on master, specifically readme. I hope you will like it.
Currently only Nitrokey Pro is supported, Nitrokey Storage should join shortly. The C API / Python API should be stable, the C++ API is subject to change.

**Comment by cornelinux** [comment]
Hi @szszszsz 
What are your recommendations on the operating system? I am running ubuntu 14.04. Obviously my gcc is too old (no c++14) and using clang-3.8 I get:

<pre>
/usr/bin/../lib/gcc/x86_64-linux-gnu/4.8/../../../../include/c++/4.8/cstdio:120:11: error: no member named 'gets' in the global namespace
  using ::gets;
        ~~^
In file included from device.cc:7:
./include/device.h:13:26: error: expected namespace name
    using namespace std::chrono_literals;
                    ~~~~~^
device.cc:88:31: error: no matching literal operator for call to 'operator""ms' with argument of type 'unsigned long long' or 'const char *',
      and no matching literal operator template
    m_send_receive_delay = 100ms;
                              ^
device.cc:97:32: error: no matching literal operator for call to 'operator""ms' with argument of type 'unsigned long long' or 'const char *',
      and no matching literal operator template
    m_send_receive_delay = 1000ms;
                               ^
4 errors generated.
</pre>

...ready to provide feedback :-)

**Comment by cornelinux** [comment]
OK, I am running on Ubuntu 16.04.
Compiled fine.

When I now try to run the python_bindings_example.py the program runs right to after the login.
Shouldn't the login ask for a PIN? It does not.

Then the get_hotp_code fails with:

    terminate called after throwing an instance of 'std::runtime_error'
      what():  Attempted HID send on an invalid descriptor.
    Aborted (core dumped)

**Comment by szszszsz** [comment]
Hi @cornelinux !
Sorry for late reply, I was offline since yesterday. 

I have updated mentioned Python example script (on `master`). It will ask for PIN and write second HOTP slot (numbered 1) with RFC test secret then print returned codes. It also writes default general config to the device. `NK_login` only connects with device (it might need name change). Not all operations needs PIN so it is not asked with login function. What is more, library does not store any passwords, so they has to be supplied on each authentication function before something requiring admin privileges will be run. The `no store` policy might be less convenient, but should be more secure.
This could be later implemented on request.

In case you would like to check all use cases please check `unittests/test_python_bindings.py`. The script can be run using `py.test`. Be aware however tests *will* overwrite whole content of the stick (`factory reset` is also tested). 

As for Ubuntu 14.04, I am not sure will installation of latest clang suffice. As far as I remember it uses GCC libraries thus it might not work properly.

If any questions arise please let me know.

**Comment by cornelinux** [comment]
Hi @szszszsz 
great. The test script now works for me. I will dive deeper and see how to integrate it into my project.
Thanks a lot!

**Comment by szszszsz** [comment]
Hi!
I hope it will work for you well!
In case if any new ideas/corrections/implementation requests please create separate issue, similarly to this. Please also do not hesitate to report any new ones - all contribution is welcomed!
